File: src/python/entityfx/writer.py

import os

class Writer(object):
    
    @property
    def use_console(self) -> bool:
        return self.__use_console

    @use_console.setter
    def use_console(self, value) -> bool:
        self.__use_console = value
        return self.__use_console
    
    @property
    def file_path(self) -> str:
        return self.__file_path
    
    def __init__(self, file_path : str=None) -> None:
        self.__output = ""
        self.__use_console = True
        self.__use_file = False
        if (file_path is not None): 
            self.__file_path = file_path
            self.__use_file = True
            self.__stream = open(file_path, "a+", 1)

        self.__is_color = True
        if os.name == "nt":
            self.__is_color = False

    @property
    def output(self) -> str:
        return str(self.__output)
    
    def write_line(self, format : str=None, *args : object) -> None:
        if format is not None:
            self.write_color("\033[1;30m", format, *args)
        if (self.use_console): 
            print('')
            self.__output += '\n'
            if (self.__use_file): 
                self.__stream.write('\n')
       
    def write_header(self, format : str, *args : object) -> None:
        self.write_color("\033[1;36m", format, *args)
        self.write_line()
       
    def write_color(self, color : str, format : str, *args : object) -> None:
        formatted = format.format(*args)
        self.__output += formatted
        if (self.use_console): 
            to_print = '{color}{formatted}\033[0m'.format(color=color, formatted=formatted) if self.__is_color else formatted
            print(to_print, end='', flush=True)
            if (self.__use_file): 
                self.__stream.write(formatted)

    def write(self, format : str, *args: object) -> None:
        self.write_color("\033[1;30m", format, *args)
    
    def write_value(self, format : str, *args: object) -> None:
        self.write_color("\033[1;32m", format, *(args))
    
    def write_title(self, format : str, *args: object) -> None:
        self.write_color("\033[1;37m", format, *(args))

File: src/go/entityfx/utils/writer.go
package utils

import (
	"fmt"
	"os"
	"runtime"
	"sync"
)

type WriterType interface {
	Write(format string, a ...interface{}) (n int, err error)

	WriteLine(format string, a ...interface{}) (n int, err error)

	WriteColor(color string, format string, a ...interface{}) (n int, err error)

	WriteNewLine() (n int, err error)

	WriteHeader(format string, a ...interface{}) (n int, err error)

	WriteValue(format string, a ...interface{}) (n int, err error)

	WriteTitle(format string, a ...interface{}) (n int, err error)

	UseConsole(value bool)

	GetOutput() string
}

type Writer struct {
	useConsole bool
	useFile    bool
	filePath   string
	output     string
	f          *os.File
	mu         *sync.Mutex
}

func NewWriter(filePath string) WriterType {
	var mutex sync.Mutex
	w := &Writer{true, false, filePath, "", nil, &mutex}

	if filePath != "" {
		w.useFile = true
		w.f, _ = os.Create(filePath)
	}

	return w
}

func (w *Writer) WriteColor(color string, format string, a ...interface{}) (n int, err error) {
	var formatted = fmt.Sprintf(format, a...)
	w.output += formatted
	if w.useConsole {
		w.mu.Lock()
		if runtime.GOOS == "windows" {
			n, err = fmt.Print(formatted)
		} else {
			n, err = fmt.Print(color + formatted + "\033[0m")
		}
		if w.useFile {
			w.f.WriteString(formatted)
		}
		w.mu.Unlock()
		return n, err
	}

	return n, err
}

func (w *Writer) UseConsole(value bool) {
	w.useConsole = value
}

func (w *Writer) GetOutput() string {
	return w.output
}

func (w *Writer) WriteNewLine() (n int, err error) {
	w.output += "\n"
	if !w.useConsole {
		return
	}
	w.mu.Lock()
	if w.useFile {
		w.f.WriteString("\n")
	}
	n, err = fmt.Println()
	w.mu.Unlock()
	return n, err
}

func (w *Writer) Write(format string, a ...interface{}) (n int, err error) {
	return w.WriteColor("\033[1;30m", format, a...)
}

func (w *Writer) WriteLine(format string, a ...interface{}) (n int, err error) {
	n, err = w.WriteColor("\033[1;30m", format, a...)
	w.WriteNewLine()
	return n, err
}

func (w *Writer) WriteHeader(format string, a ...interface{}) (n int, err error) {
	n, err = w.WriteColor("\033[1;36m", format, a...)
	w.WriteNewLine()
	return n, err
}

func (w *Writer) WriteValue(format string, a ...interface{}) (n int, err error) {
	return w.WriteColor("\033[1;32m", format, a...)
}

func (w *Writer) WriteTitle(format string, a ...interface{}) (n int, err error) {
	return w.WriteColor("\033[1;37m", format, a...)
}
